ligo/swaps.mligo

```diff
@@ -165,3 +165,33 @@ let y_to_x (s : storage) (p : y_to_x_param) : result =
         let op_receive_y = y_transfer Tezos.sender Tezos.self_address dy_spent in
         let op_send_x = x_transfer Tezos.self_address p.to_dx dx_received in
         ([op_receive_y ; op_send_x], s_new)
+
+
+(* Trade X with X' in a different contract. *)
+let x_to_x_prime (s : storage) (p : x_to_x_prime_param) : result =
+
+    (* Ensure provided contract address is correct. *)
+    let x_prime_contract =
+        match (Tezos.get_entrypoint_opt "%y_to_x" p.x_prime_contract
+            : y_to_x_param contract option) with
+        | Some contract -> contract
+        | None -> (failwith("TODO: incorrect segmented cfmm contract") : y_to_x_param contract) in
+
+    (* Swap X to Y *)
+    let ((ops, s_new), dy_received) = x_to_y s
+            ({ dx = p.dx
+             ; deadline = p.deadline
+             ; min_dy = 0n
+             ; to_dy = Tezos.self_address
+             }) in
```

Comment:
One comment here is that this should work, but it is slightly inefficient.

If you look at the implementation of `x_to_y` most of the parameters given here aren't actually useful, namely:
- `deadline` is checked, but could be avoided, because ultimately it will be checked by the following `y_to_x` call
- `min_dy` is checked, although we know it won't ever need to
- (main point) `to_dy` is used to `transfer` from `self_address` to itself, which can be avoided

So, what we actually need here is only part of `x_to_y`, which I think we should extract to a different function.

Reply:
@pasqu4le  I couldn't think of a proper name for the function to extract to. So I updated it to only use the functions that it needs:
```
    (* Swap X to Y *)
    let r = x_to_y_rec {s = s ; dx = p.dx ; dy = 0n} in
    let dx_spent = assert_nat (p.dx - r.dx, internal_309) in
    let dy_received = r.dy in
    let s_new = {s with balance = {x = s.balance.x + dx_spent ;  y = assert_nat (s.balance.y - dy_received, internal_insufficient_balance_err)}} in
    let op_receive_x =
        (* TOOD: Do we allow 0 swap? *)
        if dy_received <= 0n
        then (failwith smaller_than_min_asset_err : operation)
        else x_transfer Tezos.sender Tezos.self_address dx_spent in
```

I added a check to ensure `dy_received` is bigger than 0n, not sure if this is necessary.

On another note, I still couldn't figure out how do we ensure that the `x_prime_contract`'s `y` token is the same as the current contract. Should we add a view entrypoint to get the x and y tokens information of the contract?

Reply:
> I added a check to ensure dy_received is bigger than 0n, not sure if this is necessary.

I would imagine that we can avoid it :shrug: 
After AFAIU a `dy` of `0` will result in a number of `X'` of `0`, but we give the second contract a `min_dx`, so that's the one that should fail if `0` is not enough.

> I couldn't think of a proper name for the function to extract to

I would still make it a common function, simply because if we change one we'll need to change the other, so this is useful to have in one place. How about `update_storage_x_to_y` or something like that?

>  still couldn't figure out how do we ensure that the x_prime_contract's y token is the same as the current contract.

oh, that's a very good point. We need to check that to avoid people abusing this to exchange different tokens.
I think we should rely on the fact that `y_to_x` requires the second contract to be an `operator` (or to be `approve`d in case it's FA1.2) for the current contract.

AFAIK it should be possible for use to do:
1. set the second contract as operator of the current one
2. call `y_to_x` on the second contract, which should work now
3. unset the second contract as operator

Reply:
> AFAIK it should be possible for use to do:
> 
>     1. set the second contract as operator of the current one
> 
>     2. call `y_to_x` on the second contract, which should work now
> 
>     3. unset the second contract as operator

@pasqu4le I've made the changes according to your feedback. A few things that I assumed:
- In case of Y is FA1.2, we call `approve` instead of `update_operator`
- We are guaranteed that the operations are executed in order.

Let me know if my assumption is correct. And I also made other changes you mentioned, should be ready for re-review.

Reply:
both your assumption seem correct to me :+1: 

I saw the other changes as well and I have no more comments